### PR TITLE
Initial draft of scheduler.

### PR DESCRIPTION
**Pull request by porcuquine**
@keyvank This is still very rough and not quite complete. However, it should be enough for you to see generally how this will work if you want/need to start preparing to use. ~Still needs some outer wrapping to drive the polling and accept scheduling requests.~

TODO:
- [x] Driver loop
- [x] Preemption interface
- [x] Preemption implementation
- [x] Tests. [What I implemented is not enough to have high confidence in overall correctness. However, it is enough to show example usage and suggests that the interface and basic behavior work well enough to begin integrating and building on.]
- [ ] Thorough tests
~- [ ] (maybe) some mechanism to manage pools of static tasks, since this will be somewhat tedious and common to all tasks.~
- [x] Respect resource order preference for scheduled tasks.
- [x] Enhance test to demonstrate (possibly first ensuring) that channels can be used  to return task results.

### CODE REVIEW

File: src/scheduler.rs [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+                .or_insert(Default::default())
+                .insert(task_file);
+            self.own_tasks.insert(task_ident.clone(), Mutex::new(&task));
+        }
+        Ok(())
+    }
+}
+
+pub struct FSResourceScheduler<'a, R: Resource> {
+    root_scheduler: FSScheduler<'a, R>,
```
Comment (keyvank) [inline src/scheduler.rs:226]: Shouldn't this be a reference? Maybe: `Rc<FSScheduler<'a, R>>`
My understanding is that we will have a single root FSScheduler and a few FSResourceSchedulers (One per resource) that point to the root scheduler

File: src/scheduler.rs [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+                    // Remember this ident,
+                    None => Some((ident.clone(), 1)),
+                }
+            }
+        }
+        Ok(())
+    }
+
+    fn perform_task(&self, task: &Task<R>) -> Result<(), Error> {
+        self.lock()?;
```
[the remark this answers is not in the document]
Comment (porcuquine) [reply src/scheduler.rs:359]: Yes, good catch — thanks.

File: src/scheduler.rs (hunk 1 above)
Comment (porcuquine) [reply src/scheduler.rs]: Yes, just hadn't gotten there yet. It ended up being `Rc<RefCell<…>>`.

**Comment by porcuquine** [comment]
I think this is now feature complete and *probably* now correct to a first approximation. The test are currently broken because the first iteration relied on performance being deferred until the next poll. In order to respect resource preference, tasks are now immediately scheduled on the first available resource (checked in preference order) at the time of scheduling. This needs more sophisticated tests anyway, but now also just to demonstrate the previous degree of correctness. Needs some thought.

I also notice, though, that CI is not set up, so the test failures are invisible here. @keyvank can you please add (separate PR — we'll merge that then I will rebase this)?

**Comment by vmx** [comment]
This has been an idea a long time ago and it turned out that it wasn't needed. Hence closing this PR.

**Pull request closed without merging**